### PR TITLE
[Xamarin.Android.Build.Tasks] fix for <Aapt/>

### PR DESCRIPTION
**Pull request by jonathanpeppers**
Context: http://build.devdiv.io/2527883

Our integration build for the Xamarin designer has caught a bug in the
following situation:

* macOS-only (Windows is OK)
* `$(AndroidUseAapt2)` is `False`
* Using `Xamarin.Android.Support.CustomTabs` version 23.x

I was able to reproduce the problem by changing the
`BuildTest.ResourceExtraction` test.

The error message is:

    /Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.targets(1765,2):
        error APT0000: resource directory '/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/ADT-0/RT-MyDriving-1/MyDriving.Android/obj/Debug/resourcecache/3288F845948C6081AB114E82B328E204/res' does not exist

The reason this fails is that this particular library does not have a
`res` directory:

    $ ls -l ~/.local/share/Xamarin/Xamarin.Android.Support.CustomTabs/23.1.1.0/embedded
    total 64
     8 -rw-r--r--  1 jonathanpeppers  staff    861 Mar 26 11:18 AndroidManifest.xml
     0 drwxr-xr-x  3 jonathanpeppers  staff     96 Mar 26 11:18 aapt
    56 -rw-r--r--  1 jonathanpeppers  staff  24743 Mar 26 11:18 classes.jar

The fix is to add a `Directory.Exists` check, this seems to fix the
unit test *and* the failing project used by the designer build.

### CODE REVIEW

File: src/Xamarin.Android.Build.Tasks/Tasks/Aapt.cs [hunk 1]
```diff
@@ -331,9 +331,14 @@ protected string GenerateCommandLineCommands (string ManifestFile, string curren
 			if (AdditionalResourceDirectories != null)
 				foreach (var resdir in AdditionalResourceDirectories)
 					cmd.AppendSwitchIfNotNull ("-S ", resdir.ItemSpec.TrimEnd ('\\'));
-			if (AdditionalAndroidResourcePaths != null)
-				foreach (var dir in AdditionalAndroidResourcePaths)
-					cmd.AppendSwitchIfNotNull ("-S ", Path.Combine (dir.ItemSpec.TrimEnd (System.IO.Path.DirectorySeparatorChar), "res"));
+			if (AdditionalAndroidResourcePaths != null) {
+				foreach (var dir in AdditionalAndroidResourcePaths) {
+					var resdir = Path.Combine (dir.ItemSpec, "res");
+					if (Directory.Exists (resdir)) {
```
Comment (jonathanpeppers) [inline src/Xamarin.Android.Build.Tasks/Tasks/Aapt.cs]: @dellis1972 do you think it's worth adding a `Directory.Exists` check for every path passed via `-S`?

Comment (dellis1972) [reply src/Xamarin.Android.Build.Tasks/Tasks/Aapt.cs]: I would say yes.

File: src/Xamarin.Android.Build.Tasks/Tasks/Aapt.cs [hunk 2]
```diff
@@ -328,12 +328,22 @@ protected string GenerateCommandLineCommands (string ManifestFile, string curren
 				cmd.AppendSwitchIfNotNull ("-F ", currentResourceOutputFile + ".bk");
 			// The order of -S arguments is *important*, always make sure this one comes FIRST
 			cmd.AppendSwitchIfNotNull ("-S ", resourceDirectory.TrimEnd ('\\'));
-			if (AdditionalResourceDirectories != null)
-				foreach (var resdir in AdditionalResourceDirectories)
-					cmd.AppendSwitchIfNotNull ("-S ", resdir.ItemSpec.TrimEnd ('\\'));
-			if (AdditionalAndroidResourcePaths != null)
-				foreach (var dir in AdditionalAndroidResourcePaths)
-					cmd.AppendSwitchIfNotNull ("-S ", Path.Combine (dir.ItemSpec.TrimEnd (System.IO.Path.DirectorySeparatorChar), "res"));
+			if (AdditionalResourceDirectories != null) {
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Tasks/Aapt.cs]: Do we really need to repeat this much code, changing only the `ITaskItem[]` source?

For example, with C# local functions:

```csharp
void AddResourceDirectories(ITaskItem[] directories) {
    if (directories == null) {
        return;
    }
    foreach (var dir in directories) {
        if (!Directory.Exists (dir.ItemSpec))
            continue;
        var resdir = dir.ItemSpec.TrimEnd ('\\');
        cmd.AppendSwitchIfNotNull ("-S ", resdir);
    }
}
AddResourceDirectories (AdditionalResourceDirectories);
AddResourceDirectories (AdditionalAndroidResourcePaths);
```

Comment (jonathanpeppers) [reply src/Xamarin.Android.Build.Tasks/Tasks/Aapt.cs]: One does a `Path.Combine (dir, "res")` and the other does `TrimEnd()`.

I could use a local function, here but I'd also have to do something like:
```csharp
void AddResourceDirectories(ITaskItem[] directories, Func<string, string> func) {
    if (directories == null) {
        return;
    }
    foreach (var dir in directories) {
        if (!Directory.Exists (dir.ItemSpec))
            continue;
        var resdir = func (dir.ItemSpec);
        cmd.AppendSwitchIfNotNull ("-S ", resdir);
    }
}
AddResourceDirectories (AdditionalResourceDirectories, d => d.TrimEnd ('\\'));
AddResourceDirectories (AdditionalAndroidResourcePaths, d => Path.Combine (d, "res"));
```
At that point, not sure if it's simpler?

**Pull request merged**